File: validate_bst.java
// Time Complexity : O(n)
// Space Complexity : O(h) - height of the recursive stack
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No 


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // public boolean isValidBST(TreeNode root) {
    //     inorder(root);
    //     return true; 
    // }
    // private void inorder(TreeNode root){
    //     if(root==null) return;

    //     inorder(root.left);
    //     // st.pop() under the hood
    //     System.out.print("up");
    //     System.out.println(root.val);
    //     inorder(root.right);
    //     // st.pop() under the hood
    //     System.out.print("down");
    //     System.out.println(root.val);
    // }

    // // Using Recurssion, where there is no control using the flag
    // TreeNode prev;
    // boolean flag;
    //  public boolean isValidBST(TreeNode root) {
    //     flag = true;
    //     inorder(root);
    //     return flag; 
    // }
    // private void inorder(TreeNode root){
    //     if(root==null) return;
    //     inorder(root.left);
    //     // st.pop() under the hood
    //     if(prev!=null && prev.val>=root.val){
    //         flag = false;
    //     }
    //     prev = root;
    //     inorder(root.right);
    //     // st.pop() under the hood
    // }
    // Using Recurssion void method, where there is control using the flag
    // TreeNode prev;
    // boolean flag;
    //  public boolean isValidBST(TreeNode root) {
    //     flag = true;
    //     inorder(root);
    //     return flag; 
    // }
    // private void inorder(TreeNode root){
    //     if(root==null) return;
    //     if(flag){
    //         inorder(root.left);
    //     }
    //     // st.pop() under the hood
    //     if(prev!=null && prev.val>=root.val){
    //         flag = false;
    //     }
    //     prev = root;
    //     if(flag)
    //         inorder(root.right);
    //     // st.pop() under the hood
    // }

    // Using Recurssion boolean method, if you pass prev as local variable
    // the values will be overwritten
    TreeNode prev;
    public boolean isValidBST(TreeNode root) {
        return inorder(root);
    }
    private boolean inorder(TreeNode root){
        if(root==null) return true;
        if( inorder(root.left) ==  false) return false;
        // st.pop() under the hood
        if(prev!=null && prev.val>=root.val){
            return false;
        }
        prev = root;
        return inorder(root.right);
        // st.pop() under the hood
    }
}

File: validateBST.js
// Time Complexity : O(n)
// Space Complexity : O(n) || O(h) - height of the tree
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// optimized approach 1
// We only rely on the stack as extra space, instead of the additional array to run a loop on.
var isValidBST = function (root) {
  if (root === null) return true;
  const stack = [];
  let prev = null;

  while (root !== null || stack.length) {
    while (root !== null) {
      stack.push(root);
      root = root.left;
    }
    root = stack.pop();
    if (prev !== null && prev.val >= root.val) {
      return false;
    }
    prev = root;
    root = root.right;
  }

  return true;
};

// Optimized approach

// Additional extra space approach. Not most optimal Space Complexity.
// Also uses an extra loop, which while is dropped in asymptotic complexity
// still is a slower tc than other solutions.
var isValidBST = function (root) {
  if (root === null) return true;
  const stack = [];
  const ascArr = []; // will hold root vals in asc order

  while (root !== null || stack.length) {
    while (root !== null) {
      stack.push(root);
      root = root.left;
    }
    root = stack.pop();
    ascArr.push(root.val);
    root = root.right;
  }

  for (let i = 1; i < ascArr.length; i++) {
    if (ascArr[i] <= ascArr[i - 1]) {
      // must ensure the values are not equal. If equal, return false
      return false;
    }
  }
  return true;
};

[100, 50, 150, 20, 60, 120, 160, 15, 25, 55, 70, null, null, 159, 161];
